File: src/agent/agent-sdk/nodejs/src/dockercli.ts
/*
 * BK-CI Agent SDK - docker/podman CLI 封装
 *
 * 移植 Go agent 的 src/pkg/dockercli/dockercli.go。基于 child_process.spawn
 * 直接调用 docker/podman 命令行（非 daemon API），零第三方依赖。
 *
 * runtime 选择：环境变量 DEVOPS_AGENT_CONTAINER_RUNTIME，默认 "docker"。
 * 支持：serverOS / imageExists / pullImage(凭据 stdin login) / createContainer /
 *       startContainer / waitContainer / removeContainer / containerLogs。
 * 日志脱敏：docker -e KEY=VALUE 中 KEY 含 secret/password/token/credential 时 mask 值。
 */

import { spawn } from 'child_process';
import * as os from 'os';
import * as fs from 'fs';
import * as path from 'path';

export const DevopsAgentContainerRuntime = 'DEVOPS_AGENT_CONTAINER_RUNTIME';

export type DockerLogLevel = 'DEBUG' | 'INFO' | 'WARN' | 'ERROR';

export interface DockerLogEntry {
  level: DockerLogLevel;
  message: string;
}

export type DockerEventLogFn = (entry: DockerLogEntry) => void;

/** 读取 runtime 二进制名（对应 Go RuntimeBinary），默认 docker。 */
export function runtimeBinary(): string {
  const v = (process.env[DevopsAgentContainerRuntime] ?? '').trim();
  return v !== '' ? v : 'docker';
}

interface RunResult {
  stdout: string;
  stderr: string;
  err: Error | null;
}

const sensitiveEnvKeys = ['secret', 'password', 'token', 'credential'];

export class DockerRunner {
  private readonly workDir: string;
  private readonly binary: string;
  private readonly eventf?: DockerEventLogFn;

  constructor(workDir: string, eventf?: DockerEventLogFn, binary?: string) {
    this.workDir = workDir;
    this.binary = binary ?? runtimeBinary();
    this.eventf = eventf;
  }

  binaryName(): string {
    return this.binary;
  }

  async serverOS(): Promise<string> {
    const r = await this.run(null, ['version', '--format', '{{.Server.Os}}']);
    return r.stdout.trim();
  }

  /**
   * 镜像是否本地存在（依赖 image inspect 退出码，不解析 stderr 文本）。
   * daemon 不可达时返回 false（后续 pull/create 会暴露真实错误）。
   */
  async imageExists(image: string): Promise<boolean> {
    const r = await this.run(null, ['image', 'inspect', image]);
    return r.err === null;
  }

  /** 拉取镜像；有凭据时用临时 config 目录 + --password-stdin login。返回合并日志输出。 */
  async pullImage(image: string, user: string, password: string): Promise<string> {
    if (!user || !password) {
      const r = await this.run(null, ['pull', image]);
      if (r.err) throw r.err;
      return r.stdout + r.stderr;
    }

    const os = await import('os');
    const fs = await import('fs');
    const path = await import('path');
    const cfgDir = await fs.promises.mkdtemp(path.join(os.tmpdir(), 'bkci-docker-config-'));
    try {
      const registry = registryFromImage(image);
      const loginArgs = ['--config', cfgDir, 'login', '-u', user, '--password-stdin'];
      if (registry !== '') loginArgs.push(registry);
      const loginRes = await this.run(Buffer.from(password), loginArgs);
      if (loginRes.err) throw loginRes.err;

      const pullRes = await this.run(null, ['--config', cfgDir, 'pull', image]);
      if (pullRes.err) throw pullRes.err;
      return pullRes.stdout + pullRes.stderr;
    } finally {
      await fs.promises.rm(cfgDir, { recursive: true, force: true }).catch(() => undefined);
    }
  }

  /** 创建容器，返回 container id（取输出最后一行）。 */
  async createContainer(args: string[]): Promise<string> {
    const r = await this.run(null, ['create', ...args]);
    if (r.err) throw r.err;
    const lines = r.stdout.trim().split('\n');
    const last = (lines[lines.length - 1] ?? '').trim();
    if (last === '') {
      throw new Error(`empty container id returned from ${this.binary} create`);
    }
    return last;
  }

  async startContainer(containerId: string): Promise<void> {
    const r = await this.run(null, ['start', containerId]);
    if (r.err) throw r.err;
  }

  async stopContainer(containerId: string): Promise<void> {
    const r = await this.run(null, ['stop', '-t', '0', containerId]);
    if (r.err) throw r.err;
  }

  async removeContainer(containerId: string): Promise<void> {
    const r = await this.run(null, ['rm', '-f', containerId]);
    if (r.err) throw r.err;
  }

  /** 等待容器结束，返回退出码（对应 docker wait）。 */
  async waitContainer(containerId: string): Promise<number> {
    const r = await this.run(null, ['wait', containerId]);
    if (r.err) throw r.err;
    const code = parseInt(r.stdout.trim(), 10);
    if (Number.isNaN(code)) {
      throw new Error(`parse wait exit code failed: ${r.stdout.trim()}`);
    }
    return code;
  }

  async containerLogs(containerId: string): Promise<string> {
    const r = await this.run(null, ['logs', containerId]);
    // 与 Go 一致：即便有错也返回已收集的输出
    return r.stdout + r.stderr;
  }

  /** 底层执行：spawn runtime，收集 stdout/stderr，脱敏后按级别记日志。 */
  private run(stdin: Buffer | null, args: string[]): Promise<RunResult> {
    this.log(classifyCommandLevel(args), formatCommand(this.binary, args));
    return new Promise((resolve) => {
      const child = spawn(this.binary, args, {
        cwd: this.workDir,
        env: process.env,
      });

      let stdout = '';
      let stderr = '';
      child.stdout.on('data', (c) => (stdout += c.toString()));
      child.stderr.on('data', (c) => (stderr += c.toString()));

      const finish = (err: Error | null): void => {
        if (stdout.trim() !== '') {
          this.log(classifyStreamLevel(false, err, stdout, args), `[stdout]\n${stdout.trim()}`);
        }
        if (stderr.trim() !== '') {
          this.log(classifyStreamLevel(true, err, stderr, args), `[stderr]\n${stderr.trim()}`);
        }
        if (err) {
          resolve({
            stdout,
            stderr,
            err: new Error(`${formatCommand(this.binary, args)} failed: ${err.message}`),
          });
        } else {
          resolve({ stdout, stderr, err: null });
        }
      };

      child.on('error', (err) => finish(err));
      child.on('close', (code) => {
        finish(code === 0 ? null : new Error(`exit code ${code}`));
      });

      if (stdin !== null) {
        child.stdin.write(stdin);
      }
      child.stdin.end();
    });
  }

  private log(level: DockerLogLevel, message: string): void {
    if (this.eventf) {
      this.eventf({ level, message });
      return;
    }
    console.info(`[agent-sdk][docker][${level}]`, message);
  }
}

/** 从镜像名推断 registry host（对应 Go registryFromImage）。 */
export function registryFromImage(image: string): string {
  const s = image.replace(/^https?:\/\//, '').trim();
  const idx = s.indexOf('/');
  if (idx < 0) return '';
  const first = s.slice(0, idx);
  if (first.includes('.') || first.includes(':') || first === 'localhost') {
    return first;
  }
  return '';
}

/** 拼接命令用于日志（对含敏感 env 值脱敏）。 */
function formatCommand(binary: string, args: string[]): string {
  const parts = [binary];
  let maskNext = false;
  for (const arg of args) {
    let display = arg;
    if (maskNext) {
      display = maskEnvValue(arg);
      maskNext = false;
    } else if (arg === '-e' || arg === '--env') {
      maskNext = true;
    }
    if (/[ \t\n"']/.test(display)) {
      parts.push(JSON.stringify(display));
    } else {
      parts.push(display);
    }
  }
  return parts.join(' ');
}

function maskEnvValue(env: string): string {
  const eqIdx = env.indexOf('=');
  if (eqIdx < 0) return env;
  const key = env.slice(0, eqIdx).toLowerCase();
  for (const s of sensitiveEnvKeys) {
    if (key.includes(s)) {
      return env.slice(0, eqIdx + 1) + '******';
    }
  }
  return env;
}

function classifyCommandLevel(args: string[]): DockerLogLevel {
  if (args.length >= 2 && args[0] === 'image' && args[1] === 'inspect') {
    return 'DEBUG';
  }
  return 'INFO';
}

function classifyStreamLevel(
  isStderr: boolean,
  runErr: Error | null,
  output: string,
  args: string[]
): DockerLogLevel {
  if (runErr) {
    if (isExpectedFailure(args)) return 'INFO';
    return isStderr ? 'ERROR' : 'WARN';
  }
  if (isStderr && looksLikeWarning(output)) return 'WARN';
  return 'INFO';
}

function isExpectedFailure(args: string[]): boolean {
  return args.length >= 2 && args[0] === 'image' && args[1] === 'inspect';
}

function looksLikeWarning(output: string): boolean {
  const s = output.toLowerCase();
  return s.includes('warning') || s.includes('warn:') || s.includes('deprecated');
}

File: src/agent/agent-sdk/python/bkci_agent_sdk/docker_cli.py
"""Dependency-free Docker/Podman CLI runner."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
import logging
import os
from pathlib import Path
import shutil
import tempfile
from typing import Callable, Literal

DEVOPS_AGENT_CONTAINER_RUNTIME = "DEVOPS_AGENT_CONTAINER_RUNTIME"
DockerLogLevel = Literal["DEBUG", "INFO", "WARN", "ERROR"]


@dataclass(slots=True)
class DockerLogEntry:
    level: DockerLogLevel
    message: str


DockerEventLogFn = Callable[[DockerLogEntry], None]


@dataclass(slots=True)
class _RunResult:
    stdout: str
    stderr: str
    error: str | None


def runtime_binary() -> str:
    return os.environ.get(DEVOPS_AGENT_CONTAINER_RUNTIME, "").strip() or "docker"


class DockerRunner:
    def __init__(
        self,
        work_dir: str | Path,
        event_fn: DockerEventLogFn | None = None,
        binary: str | None = None,
    ) -> None:
        self.work_dir = str(work_dir)
        self.binary = binary or runtime_binary()
        self.event_fn = event_fn

    def binary_name(self) -> str:
        return self.binary

    async def server_os(self) -> str:
        result = await self._run(None, ["version", "--format", "{{.Server.Os}}"])
        return result.stdout.strip()

    async def image_exists(self, image: str) -> bool:
        return (await self._run(None, ["image", "inspect", image])).error is None

    async def pull_image(self, image: str, user: str, password: str) -> str:
        if not user or not password:
            result = await self._run(None, ["pull", image])
            if result.error:
                raise RuntimeError(result.error)
            return result.stdout + result.stderr
        config_dir = Path(tempfile.mkdtemp(prefix="bkci-docker-config-"))
        try:
            registry = registry_from_image(image)
            login_args = [
                "--config",
                str(config_dir),
                "login",
                "-u",
                user,
                "--password-stdin",
            ]
            if registry:
                login_args.append(registry)
            login = await self._run(password.encode(), login_args)
            if login.error:
                raise RuntimeError(login.error)
            pull = await self._run(None, ["--config", str(config_dir), "pull", image])
            if pull.error:
                raise RuntimeError(pull.error)
            return pull.stdout + pull.stderr
        finally:
            await asyncio.to_thread(shutil.rmtree, config_dir, ignore_errors=True)

    async def create_container(self, args: list[str]) -> str:
        result = await self._run(None, ["create", *args])
        if result.error:
            raise RuntimeError(result.error)
        lines = result.stdout.strip().splitlines()
        container_id = lines[-1].strip() if lines else ""
        if not container_id:
            raise RuntimeError(f"empty container id returned from {self.binary} create")
        return container_id

    async def start_container(self, container_id: str) -> None:
        result = await self._run(None, ["start", container_id])
        if result.error:
            raise RuntimeError(result.error)

    async def stop_container(self, container_id: str) -> None:
        result = await self._run(None, ["stop", "-t", "0", container_id])
        if result.error:
            raise RuntimeError(result.error)

    async def remove_container(self, container_id: str) -> None:
        result = await self._run(None, ["rm", "-f", container_id])
        if result.error:
            raise RuntimeError(result.error)

    async def wait_container(self, container_id: str) -> int:
        result = await self._run(None, ["wait", container_id])
        if result.error:
            raise RuntimeError(result.error)
        try:
            return int(result.stdout.strip())
        except ValueError as error:
            raise RuntimeError(f"parse wait exit code failed: {result.stdout.strip()}") from error

    async def container_logs(self, container_id: str) -> str:
        result = await self._run(None, ["logs", container_id])
        return result.stdout + result.stderr

    async def _run(self, stdin: bytes | None, args: list[str]) -> _RunResult:
        self._log(_classify_command_level(args), _format_command(self.binary, args))
        try:
            process = await asyncio.create_subprocess_exec(
                self.binary,
                *args,
                cwd=self.work_dir,
                env=os.environ.copy(),
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            stdout_raw, stderr_raw = await process.communicate(stdin)
            stdout = stdout_raw.decode(errors="replace")
            stderr = stderr_raw.decode(errors="replace")
            error = None if process.returncode == 0 else f"exit code {process.returncode}"
        except OSError as exception:
            stdout, stderr, error = "", "", str(exception)

        if stdout.strip():
            self._log(
                _classify_stream_level(False, error, stdout, args),
                f"[stdout]\n{stdout.strip()}",
            )
        if stderr.strip():
            self._log(
                _classify_stream_level(True, error, stderr, args),
                f"[stderr]\n{stderr.strip()}",
            )
        if error:
            error = f"{_format_command(self.binary, args)} failed: {error}"
        return _RunResult(stdout, stderr, error)

    def _log(self, level: DockerLogLevel, message: str) -> None:
        if self.event_fn:
            self.event_fn(DockerLogEntry(level, message))
        else:
            logging.getLogger("bkci_agent_sdk.docker").log(
                logging.ERROR if level == "ERROR" else logging.INFO,
                "[%s] %s",
                level,
                message,
            )


def registry_from_image(image: str) -> str:
    value = image.removeprefix("http://").removeprefix("https://").strip()
    if "/" not in value:
        return ""
    first = value.split("/", 1)[0]
    if "." in first or ":" in first or first == "localhost":
        return first
    return ""


_SENSITIVE_ENV_KEYS = ("secret", "password", "token", "credential")


def _format_command(binary: str, args: list[str]) -> str:
    parts = [binary]
    mask_next = False
    for argument in args:
        display = _mask_env_value(argument) if mask_next else argument
        mask_next = argument in {"-e", "--env"} if not mask_next else False
        if any(character in display for character in " \t\n\"'"):
            import json

            display = json.dumps(display)
        parts.append(display)
    return " ".join(parts)


def _mask_env_value(value: str) -> str:
    if "=" not in value:
        return value
    key, raw = value.split("=", 1)
    if any(sensitive in key.lower() for sensitive in _SENSITIVE_ENV_KEYS):
        return f"{key}=******"
    return f"{key}={raw}"


def _classify_command_level(args: list[str]) -> DockerLogLevel:
    return "DEBUG" if args[:2] == ["image", "inspect"] else "INFO"


def _classify_stream_level(
    is_stderr: bool, error: str | None, output: str, args: list[str]
) -> DockerLogLevel:
    if error:
        if args[:2] == ["image", "inspect"]:
            return "INFO"
        return "ERROR" if is_stderr else "WARN"
    lowered = output.lower()
    if is_stderr and any(value in lowered for value in ("warning", "warn:", "deprecated")):
        return "WARN"
    return "INFO"
